File: nice-traffic-light/python/src/nice_traffic_light/classifier.py
"""Parse the Evagene ``risk/calculate`` response into a domain ``NiceOutcome``."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class RiskCategory(str, Enum):
    NEAR_POPULATION = "near_population"
    MODERATE = "moderate"
    HIGH = "high"


@dataclass(frozen=True)
class NiceOutcome:
    counselee_name: str
    category: RiskCategory
    refer_for_genetics_assessment: bool
    triggers: tuple[str, ...]
    notes: tuple[str, ...]


class ResponseSchemaError(ValueError):
    """Raised when the Evagene response does not match the documented NICE schema."""


def classify_nice_response(payload: dict[str, Any]) -> NiceOutcome:
    """Extract a :class:`NiceOutcome` from a ``/risk/calculate`` response.

    Strict by design — a silent default here would mask a server-side
    breaking change and leave the caller reasoning over stale assumptions.
    """
    cancer_risk = _require_dict(payload, "cancer_risk")
    return NiceOutcome(
        counselee_name=_optional_str(payload, "counselee_name"),
        category=_parse_category(_require_str(cancer_risk, "nice_category")),
        refer_for_genetics_assessment=_require_bool(cancer_risk, "nice_refer_genetics"),
        triggers=_require_str_list(cancer_risk, "nice_triggers"),
        notes=_require_str_list(cancer_risk, "notes"),
    )


def _parse_category(raw: str) -> RiskCategory:
    try:
        return RiskCategory(raw)
    except ValueError as exc:
        raise ResponseSchemaError(f"Unknown NICE category: {raw!r}") from exc


def _require_dict(container: dict[str, Any], key: str) -> dict[str, Any]:
    value = container.get(key)
    if not isinstance(value, dict):
        raise ResponseSchemaError(f"Response field {key!r} is missing or not an object.")
    return value


def _require_str(container: dict[str, Any], key: str) -> str:
    value = container.get(key)
    if not isinstance(value, str):
        raise ResponseSchemaError(f"Response field {key!r} is missing or not a string.")
    return value


def _require_bool(container: dict[str, Any], key: str) -> bool:
    value = container.get(key)
    if not isinstance(value, bool):
        raise ResponseSchemaError(f"Response field {key!r} is missing or not a bool.")
    return value


def _require_str_list(container: dict[str, Any], key: str) -> tuple[str, ...]:
    value = container.get(key, [])
    if not isinstance(value, list) or not all(isinstance(item, str) for item in value):
        raise ResponseSchemaError(f"Response field {key!r} is not a list of strings.")
    return tuple(value)


def _optional_str(container: dict[str, Any], key: str) -> str:
    value = container.get(key)
    return value if isinstance(value, str) else ""

File: nice-traffic-light/dotnet/src/NiceTrafficLight/Classifier.cs
using System.Text.Json;

namespace NiceTrafficLight;

public enum RiskCategory
{
    NearPopulation,
    Moderate,
    High,
}

public sealed record NiceOutcome(
    string CounseleeName,
    RiskCategory Category,
    bool ReferForGeneticsAssessment,
    IReadOnlyList<string> Triggers,
    IReadOnlyList<string> Notes);

public sealed class ResponseSchemaException : Exception
{
    public ResponseSchemaException(string message) : base(message) { }
}

public static class NiceClassifier
{
    public static NiceOutcome Classify(JsonElement payload)
    {
        var root = RequireObject(payload, "response");
        var cancerRisk = RequireObjectField(root, "cancer_risk");

        return new NiceOutcome(
            CounseleeName: OptionalString(root, "counselee_name"),
            Category: ParseCategory(RequireStringField(cancerRisk, "nice_category")),
            ReferForGeneticsAssessment: RequireBooleanField(cancerRisk, "nice_refer_genetics"),
            Triggers: RequireStringListField(cancerRisk, "nice_triggers"),
            Notes: RequireStringListField(cancerRisk, "notes"));
    }

    private static RiskCategory ParseCategory(string raw) => raw switch
    {
        "near_population" => RiskCategory.NearPopulation,
        "moderate" => RiskCategory.Moderate,
        "high" => RiskCategory.High,
        _ => throw new ResponseSchemaException($"Unknown NICE category: '{raw}'"),
    };

    private static JsonElement RequireObject(JsonElement value, string label)
    {
        if (value.ValueKind != JsonValueKind.Object)
        {
            throw new ResponseSchemaException($"{label} is not an object");
        }
        return value;
    }

    private static JsonElement RequireObjectField(JsonElement container, string key)
    {
        if (!container.TryGetProperty(key, out var value) || value.ValueKind != JsonValueKind.Object)
        {
            throw new ResponseSchemaException($"field '{key}' is missing or not an object");
        }
        return value;
    }

    private static string RequireStringField(JsonElement container, string key)
    {
        if (!container.TryGetProperty(key, out var value) || value.ValueKind != JsonValueKind.String)
        {
            throw new ResponseSchemaException($"field '{key}' is missing or not a string");
        }
        return value.GetString() ?? throw new ResponseSchemaException($"field '{key}' is null");
    }

    private static bool RequireBooleanField(JsonElement container, string key)
    {
        if (!container.TryGetProperty(key, out var value)
            || (value.ValueKind != JsonValueKind.True && value.ValueKind != JsonValueKind.False))
        {
            throw new ResponseSchemaException($"field '{key}' is missing or not a boolean");
        }
        return value.GetBoolean();
    }

    private static IReadOnlyList<string> RequireStringListField(JsonElement container, string key)
    {
        if (!container.TryGetProperty(key, out var value))
        {
            return Array.Empty<string>();
        }
        if (value.ValueKind != JsonValueKind.Array)
        {
            throw new ResponseSchemaException($"field '{key}' is not a list of strings");
        }

        var list = new List<string>(value.GetArrayLength());
        foreach (var item in value.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.String)
            {
                throw new ResponseSchemaException($"field '{key}' is not a list of strings");
            }
            list.Add(item.GetString() ?? string.Empty);
        }
        return list;
    }

    private static string OptionalString(JsonElement container, string key)
    {
        if (!container.TryGetProperty(key, out var value) || value.ValueKind != JsonValueKind.String)
        {
            return string.Empty;
        }
        return value.GetString() ?? string.Empty;
    }
}
